packages/type-utils/src/typeOrValueSpecifiers/typeDeclaredInPackageDeclarationFile.ts

```diff
@@ -33,12 +33,16 @@ function typeDeclaredInDeclarationFile(
   // Handle scoped packages: if the name starts with @, remove it and replace / with __
   const typesPackageName = packageName.replace(/^@([^/]+)\//, '$1__');
 
-  const matcher = new RegExp(`${packageName}|${typesPackageName}`);
+  const packageNameMatcher = new RegExp(`${packageName}|${typesPackageName}`);
+  const fileNameMatcher = new RegExp(
+    `node_modules/(?:${packageName}|@types/${typesPackageName})/`,
+  );
```

Comment:
https://www.typescriptlang.org/tsconfig/#typeRoots - if someone customizes this, then `fileNameMatcher` will be wrong. This'll need to reflect the project's `compilerOptions.typeRoots`.

Reply:
(moved to https://github.com/typescript-eslint/typescript-eslint/pull/10097/files#r1854146022 to be closer to implementation)